### PR TITLE
Automatically provide tangents

### PR DESCRIPTION
**Pull request by oxinabox**
Closes https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/67
Does option 2D witth `x ⊢ ẋ`
that is `\vdash` + <kbd>tab</kbd>
We can argue about symbols though.

- [x] input tangent
- [x] should make frule take an output tangent #66
- [x] should make frule and rrule's output tangent be auto-generated  
- [x] Should probably also do renaming from #96 to `test_frule` and `test_rrule`. Which should make deprecations easier
- [x] tests

#### before:

```julia
x = rand(4, 5)
x̄ = rand(4, 5)

ȳ = rand(2, 10)

rrule_test(reshape, ȳ, (x, x̄), ((2, 10), nothing))
rrule_test(reshape, ȳ, (x, x̄), (2, nothing), (10, nothing))
```

### After
No point declaring `x` in advance any-more.
Can just put the `rand` in-place

```julia
test_rrule(reshape, rand(4, 5), (2, 10) ⊢  nothing)
test_rrule(reshape, rand(4, 5), 2  ⊢  nothing, 10 ⊢  nothing)
```

### CODE REVIEW

**Comment by oxinabox** [comment]
I have not added tests for this explictly, but some places we still use this because we explictly want to pass special tangents in.
So we basically have real world based tests that this works.
So i think this is good.
but review should comment if they disagree

**Comment by oxinabox** [comment]
I would say that we should wait til ChainRules is updated to use this, before merging.
But it is going to be a big job to update all the tests. Probably multiple PRs.
So I think we should merge this, and then patch it if we find something was missed.

This is nonbreaking as the old way is deprecated.

File: src/generate_tangent.jl
Comment (willtebbutt) [inline src/generate_tangent.jl:28]: Could you add instructions on how to get this symbol to this docstring?

File: src/ChainRulesTestUtils.jl
Comment (oxinabox) [inline src/ChainRulesTestUtils.jl:16]: Don't need to export `Auto()` as I made the output tangent into a kwarg, so user doesn't need to specify to do it Auto, they just need to not pass that kwarg

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:62]: This should use the new `test_rrule`

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:49]: This should use new function

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:41]: This should use new function

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:149]: Should this be `output_cotangent`?

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:173]: It's a bit stilly to be attaching a tangent to the primal explicitly, then pealing it off, but it makes both given and  `Auto()` work

**Comment by oxinabox** [comment]
Ah, I forgot to push some changes.
Will do so tomorrow

**Comment by mzgubic** [comment]
That's wonderful! Please ping me when it is done, happy to review

File: src/ChainRulesTestUtils.jl
Comment (oxinabox) [reply src/ChainRulesTestUtils.jl:16]:
```suggestion
export ⊢
```

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:98]:
```suggestion
    _ensure_not_running_on_functor(f, "test_frule")
```

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:158]:
```suggestion
    _ensure_not_running_on_functor(f, "test_rrule")
```

File: test/deprecated.jl
Comment (oxinabox) [inline test/deprecated.jl:27]:
```suggestion
```

File: test/deprecated.jl
Comment (oxinabox) [inline test/deprecated.jl:26]: This is a subset of the content that was in `testers.jl` using the old API

**Review by mzgubic — COMMENTED** [review]
I love this change, makes the API so much easier to use. Should we also update the examples in the `docs` to reflect the new usage?

File: src/ChainRulesTestUtils.jl
Comment (oxinabox) [reply src/ChainRulesTestUtils.jl:15]: Yes `@deprecate` exports unless you pass the optional 3rd argument as `false`
which is handy as it means you only need to delete it in once place

File: src/deprecated.jl
Comment (oxinabox) [reply src/deprecated.jl:34]: to see deprecation warnings in Julia 1.5+ you need to start julia with `julia --depwarn=yes`.
(By default `]test` sets this flag.)

File: src/testers.jl
Comment (oxinabox) [reply src/testers.jl:69]: They normally are, I think.
but this one was already over-width; so I i just matched what was there
I am tempted to say we should just delete the listing of keyward arguments and just let them be listed in the text below.

File: src/testers.jl
Comment (oxinabox) [reply src/testers.jl:90]: we do for purposes of testing `InplaceThunks`
It is issue #66

File: src/testers.jl
Comment (mzgubic) [reply src/testers.jl:69]: thanks, sounds good

File: test/deprecated.jl
Comment (oxinabox) [reply test/deprecated.jl:26]: yes, just moved and subset,
no changes

File: src/testers.jl
Comment (mzgubic) [reply src/testers.jl:90]: It doesn't seem to be used in the body of the function for now? Is this meant to be added in a future PR?

File: src/testers.jl
Comment (oxinabox) [reply src/testers.jl:174]: the alternative is:
```suggestion
    ȳ = output_tangent === Auto() ? rand_tangent(primal) : output_tangent
```

Which maybe is cleaner?

idk yesterday I was hating on branches.
I started to write `auto_tangent(primal, output_tangent)` and realized that it overlapped with `auto_primal_and_tangent`

File: src/testers.jl
Comment (oxinabox) [reply src/testers.jl:90]: It should be, let me check

Comment (oxinabox) [reply src/testers.jl:90]: It isn't.
Adding to https://github.com/JuliaDiff/ChainRulesTestUtils.jl/pull/116/files/3644e23d3bf5ba775bb432a0abe808898e7b2834#diff-988b853ad3c22aebac9421d3a31e60f9330789df43bccfb9fb75365048398fbaR121

File: docs/src/index.md
Comment (mzgubic) [inline docs/src/index.md:125]:
```suggestion
A special case of this is that if you specify it as `x ⊢ nothing` then finite differencing will not be used on that input.
```

File: docs/src/index.md
Comment (mzgubic) [inline docs/src/index.md:130]:
```suggestion
Which is correct e.g. when the `Int` represents a discrete integer like in indexing.
```

File: docs/src/index.md
Comment (mzgubic) [inline docs/src/index.md:131]:
```suggestion
But if you are testing something where the `Int` is actually a special case of a real number, then you would want to specify the tangent as a `Float64`.
```

File: docs/src/index.md
Comment (mzgubic) [inline docs/src/index.md:134]:
```suggestion
In particular, when specifying the input tangents to [`test_frule`](@ref) and the output tangent to [`test_rrule`](@ref).
```

**Review by mzgubic — APPROVED** [review]

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:119]: @mzgubic fixed

Comment (mzgubic) [reply src/testers.jl:119]: yeah thanks, saw it - lgtm

File: src/testers.jl
Comment (nickrobinson251) [reply src/testers.jl:170]: `ȳ = output_tangent === Auto() ? rand_tangent(primal) : output_tangent` is certainly much clearer, to my eyes

**Comment by CarloLucibello** [comment]
I just got the deprecation message
```
┌ Warning: `rrule_test(f, ȳ, inputs::Tuple{Any, Any}...; kwargs...)` is deprecated, use `test_rrule(f, (x ⊢ dx for (x, dx) = inputs)...; output_tangent = ȳ, kwargs...)` instead.
```
and it looked a bit too exotic. Why not pairs `=>` instead of vdash?

File: src/testers.jl
Comment (CarloLucibello) [inline src/testers.jl:130]: ẋ should be x̄. 
Maybe it should be explained also here how to produce a vdash

**Comment by oxinabox** [comment]
@CarloLucibello note the deprecations warning is misleading as to the correct action.
Correct is to just delete the partials and let the automatic process kick in.

**Comment by nickrobinson251** [comment]
> @CarloLucibello note the deprecations warning is misleading as to the correct action.
> Correct is to just delete the partials and let the automatic process kick in.

Maybe manually define the deprecation, with a `depwarn` giving the better instructions, rather than rely on `@deprecate`

Otherwise people will do what the deprecation warning tells them to

**Comment by oxinabox** [comment]
Good idea, I will see if I find time

**Pull request merged**
